buildkite/bazelci.py

```diff
@@ -3234,35 +3234,37 @@ def publish_binaries():
     bazel_hashes, bazel_nojdk_hashes = upload_bazel_binaries()
 
     # Try to update the info.json with data about our build. This will fail (expectedly) if we're
-    # not the latest build.
-    for _ in range(5):
-        latest_generation, latest_build_number = latest_generation_and_build_number()
+    # not the latest build. Only do this if we're building binaries from the main branch to avoid
+    # accidentally publishing a custom debug build as the "latest" Bazel binary.
+    if os.getenv("BUILDKITE_BRANCH") in ("master", "stable", "main"):
+        for _ in range(5):
+            latest_generation, latest_build_number = latest_generation_and_build_number()
+
+            if current_build_number <= latest_build_number:
+                eprint(
+                    (
+                        "Current build '{0}' is not newer than latest published '{1}'. "
+                        + "Skipping publishing of binaries."
+                    ).format(current_build_number, latest_build_number)
+                )
+                break
+
+            try:
+                try_publish_binaries(
+                    bazel_hashes, bazel_nojdk_hashes, current_build_number, latest_generation
+                )
+            except BinaryUploadRaceException:
+                # Retry.
+                continue
```

Comment:
That's a horrible way how GitHub prints this diff, because I only added one line and indented the rest...

If you use this option it's easier to see what changed:
![image](https://user-images.githubusercontent.com/504652/121529126-621de500-c9fc-11eb-8420-e7fc682e35e5.png)